File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.java
class Solution {
    // int mp[];
    public int numMatchingSubseq(String s, String[] words) {
        Map<String,Boolean> map = new HashMap<>();
        map.put(s,true);
        // mp = new int[26];
        int count = 0;
        // for( char c:s.toCharArray() )
        //     mp[c-'a']++;
        for( String word : words ){
            if( word.length() > s.length() )
                continue;
            else{
                if( isSub(s,word,map) )
                    count++;
            }
        }
        return count;
    }
    private boolean isSub(String s , String word, Map<String,Boolean> map){
        if( map.containsKey(word) )
            return map.get(word);
        int j = 0;
        for( int i = 0 ; i < s.length() ; i++ ){
            // if( mp[word.charAt(j) - 'a' ] == 0 )
            //     return false;
            if( s.charAt(i) == word.charAt(j) ){
                j++;
                if( j == word.length() ){
                    map.put(word,true);
                    return true;
                } 
            }
        }
        map.put(word,false);
        return false;
    }
}

File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        
        dp = {s:True}
        
        def isSub(s,word):
            if word in dp:
                return dp[word]
            j = 0
            for i in range(len(s)):
                if s[i] == word[j]:
                    j += 1
                    if j == len(word):
                        dp[word] = True
                        return True
            dp[word] = False
            return False
        
        ans = 0
        for word in words:
            if len(word) > len(s):
                continue
            else:
                if isSub(s,word):
                    ans += 1
        return ans